core/src/apps/zcash/sig_hasher.py

```diff
@@ -0,0 +1,187 @@
+from trezor.crypto.hashlib import blake2b
+from trezor.utils import HashWriter, empty_bytearray
+from apps.bitcoin.writers import (
+    write_bytes_fixed,
+    write_bytes_reversed,
+    write_tx_output,
+    write_uint32,
+    write_uint64,
+    write_bytes_prefixed,
+    write_uint8,
+    TX_HASH_SIZE,
+)
+from apps.bitcoin import scripts
+from apps.bitcoin.common import SigHashType
+
+from typing import TYPE_CHECKING
+if TYPE_CHECKING:
+    from trezor.messages import TxInput, TxOutput, SignTx, PrevTx
+    from trezor.utils import Writer
+    from apps.common.coininfo import CoinInfo
+    from typing import Sequence
+
+
+def write_hash(w: Writer, hash: bytes):
+    write_bytes_fixed(w, hash, TX_HASH_SIZE)
+
+
+def write_prevout(w: Writer, txi: TxInput):
+    write_bytes_reversed(w, txi.prev_hash, TX_HASH_SIZE)
+    write_uint32(w, txi.prev_index)
+
+
+class ZcashSigHasher:
+    def __init__(self):
+        self.header = HeaderHasher()
+        self.transparent = TransparentHasher()
+        self.sapling = SaplingHasher()
+        self.orchard = OrchardHasher()
+
+        self.tx_hash_person = None
+        self.initialized = False
+
+    def initialize(self, tx: SignTx | PrevTx):
+        """Initialize ZcashHasher with a transaction data."""
+        self.header.initialize(tx)
+
+        tx_hash_person = empty_bytearray(16)
+        write_bytes_fixed(tx_hash_person, b'ZcashTxHash_', 12)
+        write_uint32(tx_hash_person, tx.branch_id)
+        self.tx_hash_person = bytes(tx_hash_person)
+
+        self.initialized = True
+
+    def txid_digest(self):
+        """Returns the transaction identifier."""
+        assert self.initialized
+
+        h = HashWriter(blake2b(outlen=32, personal=self.tx_hash_person))
+
+        write_hash(h, self.header.digest())       # T.1: header_digest       (32-byte hash output)
+        write_hash(h, self.transparent.digest())  # T.2: transparent_digest  (32-byte hash output)
+        write_hash(h, self.sapling.digest())      # T.3: sapling_digest      (32-byte hash output)
+        write_hash(h, self.orchard.digest())      # T.4: orchard_digest      (32-byte hash output)
+
+        return h.get_digest()
```

Comment:
Note that this is dead code. We actually don't use or need `txid_digest()` anywhere, so I am wondering whether to keep the code for it. The only place that I can think of where it would be used are replacement transactions, which we do not support for Zcash, because fee bumping is not an issue. Is there any other possible use?

Reply:
I can imagine that at some point in the future we will want to display the TXID on the Trezor screen at the end of the signing process. So I don't mind keeping the code for future use, just wanted to open a discussion.

Reply:
I will keep it for future use cases and for now I will add a comment that this is a dead code.

Reply:
https://github.com/trezor/trezor-firmware/pull/2166/commits/079f0a118983a7fbecc12686c7b78ade78d2f650